hadoop-common-project/hadoop-auth/pom.xml

```diff
@@ -234,6 +235,24 @@
           <excludeFilterFile>${basedir}/dev-support/findbugsExcludeFile.xml</excludeFilterFile>
         </configuration>
       </plugin>
+      <plugin>
+        <groupId>com.google.code.maven-replacer-plugin</groupId>
+        <artifactId>replacer</artifactId>
+        <executions>
+          <execution>
+            <id>replace-sources</id>
+            <configuration>
+              <skip>false</skip>
+            </configuration>
+          </execution>
+          <execution>
+            <id>replace-test-sources</id>
+            <configuration>
+              <skip>false</skip>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
```

Comment:
Instead of replacing guava references along with protobuf references, create a separate replacer-execution in hadoop-project/pom.xml and enable by default for all projects.
```
            <execution>
              <id>replace-guava</id>
              <phase>process-sources</phase>
              <goals>
                <goal>replace</goal>
              </goals>
              <configuration>
                <skip>false</skip><!--This will run for all modules-->
                <basedir>${basedir}</basedir>
                <includes>
                  <include>src/main/java/**/*.java</include>
                  <include>src/test/java/**/*.java</include>
                </includes>
                <replacements>
                  <replacement>
                    <token>([^\.])com.google.common</token>
                    <value>$1${hadoop-thirdparty-shaded-guava-prefix}</value>
                  </replacement>
                </replacements>
              </configuration>
            </execution>
```

2. Exclude guava-replacement in hadoop-yarn-csi module. yarn-csi seems to have separate classpath, and it explicitly uses guava-20. This will not affect any downstreams.

3. Replaces all dependencies on guava to hadoop-shaded-guava (except yarn-csi)

4. In hadoop-common/pom.xml, Keep both guava (11.0.2) and hadoop-shaded-guava as dependencies. Of course code should use shaded version of guava. version 11.0.2 is used in curator libs during runtime.

Reply:
A few questions so I can better understand the suggested approach:

So source would continue to refer to unshaded guava? (unshaded guava 11?)

Rather than rewrite once as this PR does, we'd rewrite on every build? How long does the rewrite take?

hadoop-yarn-csi is bound to guava-20? Why is that? The grpc used? Interested in how a guava 20 in classpath won't mess up downstreamers who want to use something newer (or older).

For 4., hopefully we can (later) add an enforcer so non-shaded references get flagged at build time. Rather than have hadoop depend on something it doesn't use, hopefully we can work on fixing the curator dependency to fix its pom?

Thanks @vinayakumarb

Reply:
> A few questions so I can better understand the suggested approach:
> 
> So source would continue to refer to unshaded guava? (unshaded guava 11?)
Hadoop source uses shaded guava itself from hadoop-thirdparty.
> 
> Rather than rewrite once as this PR does, we'd rewrite on every build? How long does the rewrite take?
> 
This replacer is basically to make the Job easy for this PR.  Since we commit the replaced code itself, further replaces wont happen on same file. It will help to catch further references to unshaded guava. Rewrite doesnot take much time, it finishes in fraction of second. 

> hadoop-yarn-csi is bound to guava-20? Why is that? The grpc used? Interested in how a guava 20 in classpath won't mess up downstreamers who want to use something newer (or older).

yes, yarn-csi uses grpc. And grpc refers to guava-20 and it uses protobuf 3.x even before protobuf was upgraded for entire hadoop. It keeps separate directory (share/hadoop/yarn/csi/lib) for its dependencies and existing scripts doesnt add this directory into classpath of other processes.

> 
> For 4., hopefully we can (later) add an enforcer so non-shaded references get flagged at build time. Rather than have hadoop depend on something it doesn't use, hopefully we can work on fixing the curator dependency to fix its pom?
> 
yes, If we could add this enforcer, may be we need not find and replace unshaded usage of guava on every build.
For curator, curator community is not ready to shade 3 classes of guava as they are part of their API. These classes seems to be pretty stable across versions. So curator does work with guava-11 in classpath.
If we want to fix this and completely remove hadoop's dependency on guava, we may need to shade curator as well in hadoop-thirdparty and use that in hadoop.
We are fortunate enough that, these dependencies are not part of any exposed APIs.

> Thanks @vinayakumarb

Thanks @saintstack

Reply:
Thanx @vinayakumarb and @saintstack 
Since the discussion is going on here, I thought to get everything here only.
Regarding guava revert to 11, I think, as you mentioned that would be incompatible, and downstream would be excluding them only, when adding hadoop-jars as dependency. So, In the present state there won't be any problem with the version, since the hadoop jars aren't using it, and for the other dependencies they are not getting affected by this version, as tests passed both with 27 and 11. So, I think we can keep guava version as is and let downstream exclude while including hadoop-jars(most projects already do that) and run happily with there own guava versions.

Since reverting back would be problem for those, who don't have there own guava and rely on hadoop to provide guava, there code might break. 

So, I think we can keep the reversal separate from shading and if the community is OK with going back, we will revert back in a separate jira

Reply:
@vinayakumarb 

>  Since we commit the replaced code itself...

Good.

Ok on the isolation of grpc and dependencis (Can later try shading grpc later if a problem)

Thanks for update on curator (shading curator would be a pain. It does shading itself IIRC). Hive also has guava 11 in its API apparently.

> We are fortunate enough that, these dependencies are not part of any exposed APIs.
Amen

@ayushtkn 

> So, I think we can keep guava version as is ...

At 27? (Or at 11). Sorry, had trouble grokking here.

> Since reverting back would be problem for those, who don't have there own guava and rely on hadoop to provide guava, there code might break.

I think this would be ok -- i.e. breakage -- especially in a new minor version, 3.4.0. The fix is easy enough for the downstream, just add guava dependency.

> 
So, I think we can keep the reversal separate from shading and if the community is OK with going back, we will revert back in a separate jira

Sounds good.

Reply:
@ayushtkn et al.

I recently did similar work as this PR on top of the 3.3.0 release, using the exact same approach with the replacer plugin and renaming the guava namespace to the Hadoop thirdparty shaded version throughout the whole codebase. The short term goal was to solve the Hive incompatibility issue with Spark. The change worked, the Hive unit tests within the Spark codebase largely pass, no more linkage error.  However, a guava dependency still needed to be declared in the Hadoop pom.xml to overcome an enforcerer error about conflicting Guava versions that show up as transitive dependencies.  

Hope this is helpful information. If there is interest in the community, I make a PR or otherwise make that available for testing?

Reply:
It's already there in the hadoop parent pom -
https://github.com/apache/hadoop/blob/trunk/hadoop-project/pom.xml#L693

I think you are getting conflicting dependencies, while building your project? you need to add guava in your parent pom. 
Anyway you can try excluding guava as well while adding the hadoop dependency?